models/src/main/scala/coop/rchain/casper/protocol/CasperMessage.scala

```diff
@@ -0,0 +1,438 @@
+    for {
+      candidateProto <- ba.candidate.toRight("Candidate not available")
+      candidate      <- ApprovedBlockCandidate.from(candidateProto)
+      sig            <- ba.sig.toRight("Sig not available")
+    } yield BlockApproval(candidate, sig)
+}
+
+final case class ApprovedBlock(candidate: ApprovedBlockCandidate, sigs: List[Signature])
+    extends CasperMessage {
+  def toProto: ApprovedBlockProto =
+    ApprovedBlockProto()
+      .withCandidate(candidate.toProto)
+      .withSigs(sigs)
+}
+
+object ApprovedBlock {
+  def from(ba: ApprovedBlockProto): Either[String, ApprovedBlock] =
+    for {
+      candidateProto <- ba.candidate.toRight("Candidate not available")
+      candidate      <- ApprovedBlockCandidate.from(candidateProto)
+    } yield ApprovedBlock(candidate, ba.sigs.toList)
+}
+
+final case class NoApprovedBlockAvailable(identifier: String, nodeIdentifer: String)
+    extends CasperMessage {
+  def toProto: NoApprovedBlockAvailableProto =
+    NoApprovedBlockAvailableProto()
+      .withIdentifier(identifier)
+      .withNodeIdentifer(nodeIdentifer)
+}
+
+object NoApprovedBlockAvailable {
+  def from(naba: NoApprovedBlockAvailableProto): NoApprovedBlockAvailable =
+    NoApprovedBlockAvailable(naba.identifier, naba.nodeIdentifer)
+}
+
+final case class ApprovedBlockRequest(identifier: String) extends CasperMessage {
+  def toProto: ApprovedBlockRequestProto = ApprovedBlockRequestProto(identifier)
+}
+
+object ApprovedBlockRequest {
+  def from(abr: ApprovedBlockRequestProto): ApprovedBlockRequest =
+    ApprovedBlockRequest(abr.identifier)
+}
+
+final case class BlockMessage(
+    blockHash: ByteString,
+    header: Header,
+    body: Body,
+    justifications: List[Justification],
+    sender: ByteString,
+    seqNum: Int,
+    sig: ByteString,
+    sigAlgorithm: String,
+    shardId: String,
+    extraBytes: ByteString = ByteString.EMPTY
+) extends CasperMessage {
+  def toProto: BlockMessageProto = BlockMessage.toProto(this)
+}
+
+object BlockMessage {
+
+  def from(bm: BlockMessageProto): Either[String, BlockMessage] =
+    for {
+      header <- bm.header.toRight("Header not available").map(Header.from)
+      body   <- bm.body.toRight("Body not available") >>= (Body.from)
+    } yield BlockMessage(
+      bm.blockHash,
+      header,
+      body,
+      bm.justifications.toList.map(Justification.from),
+      bm.sender,
+      bm.seqNum,
+      bm.sig,
+      bm.sigAlgorithm,
+      bm.shardId,
+      bm.extraBytes
+    )
+
+  def toProto(bm: BlockMessage): BlockMessageProto =
+    BlockMessageProto()
+      .withBlockHash(bm.blockHash)
+      .withHeader(Header.toProto(bm.header))
+      .withBody(Body.toProto(bm.body))
+      .withJustifications(bm.justifications.map(Justification.toProto))
+      .withSender(bm.sender)
+      .withSeqNum(bm.seqNum)
+      .withSig(bm.sig)
+      .withSigAlgorithm(bm.sigAlgorithm)
+      .withShardId(bm.shardId)
+      .withExtraBytes(bm.extraBytes)
```

Comment:
sadly, there's no way to use the type system that no fields are omitted here.

Please add a generative roundTripTest similar to https://github.com/rchain/rchain/blob/d8f1e7af888a5245d077b5efae5a5f0505537d19/models/src/test/scala/coop/rchain/models/RhoTypesTest.scala#L19-L27

Please disregard the stacksafety tests that are piggy-backed in there :sweat_smile:

Reply:
Can you elaborate on the above?

Reply:
I was going to say:

> If I add a field to BlockMessage, there's nothing that will prevent me from not adding it to BlockMessageProto, and the related to/from methods (not the tests, not the compiler

it is though true that I'll need to provide a value out of thin air in `fromProto` to make stuff compile, at which point I'll remember about the Proto.

Still, nothing prevents me from making a typo/copy-paste error and binding two fields of BlockMessage to the same field of BlockMessageProto (making the newly-added field in the Proto unused).

A test that does:
```scala
forall(bm: BlockMessage) {
  BlockMessage.fromProto(bm.toProto) == Right(bm)
}
```
would prevent that, and similar errors (e.g. hardcoding a value in fromProto).

I'm fine if this is ticketed as a separate PR, as I later realized it requires making shapeless spit out the Arbitrary instances for the new ADT, and that might mandate a migration to scalacheck-magnolia from (presumably slower) scalacheck-shapeless.

Reply:
Oh, ok. Yeah, I was thinking about something like that. Have it on my TODO